### PR TITLE
Replicas as mandatory property

### PR DESCRIPTION
**Pull request by jkfran**

### CODE REVIEW

File: create-project [hunk 1]
```diff
@@ -38,8 +38,11 @@ def create_site(domain):
         "image": "prod-comms.docker-registry.canonical.com/" + name,
     }
 
+    data["production"] = {}
+    data["production"]["replicas"] = 5
```
Comment (nottrobin) [inline create-project]: If it's mandatory, don't we explicitly *not* want to set it here?

Comment (nottrobin) [reply create-project]: Oh no I misunderstood. This is the generator?

Comment (jkfran) [reply create-project]: What do you mean? This script could be used by other people to set up a new project, right? If they use this script we need to provide all the needed the default configuration for a new project and then he will check if he need to modify the output.

File: sites/etclite.io.yaml [hunk 2]
```diff
@@ -8,6 +8,7 @@ env:
 
 # Overrides for production
 production:
+  replicas: 2
```
Comment (nottrobin) [inline sites/etclite.io.yaml]: Shouldn't this be 5 and 3?

File: sites/staging-api.snapcraft.io.yaml [hunk 3]
```diff
@@ -53,5 +53,6 @@ nginxConfigurationSnippet: |
   more_set_headers "Link: <https://assets.ubuntu.com>; rel=preconnect; crossorigin, <https://assets.ubuntu.com>; rel=preconnect";
```
Comment (nottrobin) [inline sites/staging-api.snapcraft.io.yaml]: Missing `production` replicas here aren't we?

Comment (nottrobin) [reply sites/staging-api.snapcraft.io.yaml]:
``` bash
 $ ./konf.py production sites/staging-api.snapcraft.io.yaml 
Traceback (most recent call last):
  File "./konf.py", line 157, in <module>
    print(projectConfig.render())
  File "./konf.py", line 121, in render
    deployment_env=self.deployment_env,
  File "/home/robin/Projects/deployment-configs/env3/lib/python3.6/site-packages/jinja2/asyncsupport.py", line 76, in render
    return original_render(self, *args, **kwargs)
  File "/home/robin/Projects/deployment-configs/env3/lib/python3.6/site-packages/jinja2/environment.py", line 1008, in render
    return self.environment.handle_exception(exc_info, True)
  File "/home/robin/Projects/deployment-configs/env3/lib/python3.6/site-packages/jinja2/environment.py", line 780, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "/home/robin/Projects/deployment-configs/env3/lib/python3.6/site-packages/jinja2/_compat.py", line 37, in reraise
    raise value.with_traceback(tb)
  File "./templates/site.yaml", line 15, in top-level template code
    {% include 'deployment.yaml' %}
  File "./templates/deployment.yaml", line 21, in top-level template code
    containerPort: {{ data.containerPort | default(80) }}
jinja2.exceptions.UndefinedError: 'dict object' has no attribute 'replicas'
```

File: sites/etclite.io.yaml (hunk 2 above)
Comment (jkfran) [reply sites/etclite.io.yaml]: My bad

**Comment by nottrobin** [comment]
Another error:

``` bash
$ ./konf.py production sites/webteam-ircbot.canonical.com.yaml 
Traceback (most recent call last):
  File "./konf.py", line 157, in <module>
    print(projectConfig.render())
  File "./konf.py", line 121, in render
    deployment_env=self.deployment_env,
  File "/home/robin/Projects/deployment-configs/env3/lib/python3.6/site-packages/jinja2/asyncsupport.py", line 76, in render
    return original_render(self, *args, **kwargs)
  File "/home/robin/Projects/deployment-configs/env3/lib/python3.6/site-packages/jinja2/environment.py", line 1008, in render
    return self.environment.handle_exception(exc_info, True)
  File "/home/robin/Projects/deployment-configs/env3/lib/python3.6/site-packages/jinja2/environment.py", line 780, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "/home/robin/Projects/deployment-configs/env3/lib/python3.6/site-packages/jinja2/_compat.py", line 37, in reraise
    raise value.with_traceback(tb)
  File "./templates/site.yaml", line 29, in top-level template code
    {% include 'ingress.yaml' %}
  File "./templates/ingress.yaml", line 10, in top-level template code
    {%- if data.nginxConfigurationSnippet %}
jinja2.exceptions.UndefinedError: 'dict object' has no attribute 'nginxConfigurationSnippet'
```

File: sites/staging-api.snapcraft.io.yaml (hunk 3 above)
Comment (jkfran) [reply sites/staging-api.snapcraft.io.yaml]: But this project only exits in the staging namespace. We don't need to generate the production output, right?

**Comment by nottrobin** [comment]
Errors:

``` bash
s (replicas-staging) $ for file in $(ls sites); do echo $file >>/dev/stderr; ./konf.py production sites/$file; done > prod-output.yaml
360.canonical.com.yaml
canonical.com.yaml
certification.ubuntu.com.yaml
charmed-osm.com.yaml
cloud-init.io.yaml
cn.ubuntu.com.yaml
conjure-up.io.yaml
design.ubuntu.com.yaml
docs.conjure-up.io.yaml
docs.ubuntu.com.yaml
Traceback (most recent call last):
  File "./konf.py", line 157, in <module>
    print(projectConfig.render())
  File "./konf.py", line 121, in render
    deployment_env=self.deployment_env,
  File "/home/robin/Projects/deployment-configs/env3/lib/python3.6/site-packages/jinja2/asyncsupport.py", line 76, in render
    return original_render(self, *args, **kwargs)
  File "/home/robin/Projects/deployment-configs/env3/lib/python3.6/site-packages/jinja2/environment.py", line 1008, in render
    return self.environment.handle_exception(exc_info, True)
  File "/home/robin/Projects/deployment-configs/env3/lib/python3.6/site-packages/jinja2/environment.py", line 780, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "/home/robin/Projects/deployment-configs/env3/lib/python3.6/site-packages/jinja2/_compat.py", line 37, in reraise
    raise value.with_traceback(tb)
  File "./templates/site.yaml", line 24, in top-level template code
    {%- include 'deployment.yaml' %}
  File "./templates/deployment.yaml", line 21, in top-level template code
    containerPort: {{ data.containerPort | default(80) }}
jinja2.exceptions.UndefinedError: 'dict object' has no attribute 'replicas'
docs.vanillaframework.io.yaml
dqlite.io.yaml
engage.canonical.com.yaml
etclite.io.yaml
jaas.ai.yaml
jp.ubuntu.com.yaml
jujucharms.com.yaml
juju.is.yaml
kubeflow-news.com.yaml
limenet.snapcraft.io.yaml
maas.io.yaml
manager.assets.ubuntu.com.yaml
microk8s.io.yaml
microstack.run.yaml
mir-server.io.yaml
multipass.run.yaml
netplan.io.yaml
old-docs.jujucharms.com.yaml
old-docs.maas.io.yaml
README.md
Traceback (most recent call last):
  File "./konf.py", line 156, in <module>
    projectConfig = Konf(**vars(args))
  File "./konf.py", line 64, in __init__
    self.load_values(values_file, local_qa, docker_tag)
  File "./konf.py", line 76, in load_values
    self.values = yaml.load(values_file, Loader=yaml.FullLoader)
  File "/home/robin/Projects/deployment-configs/env3/lib/python3.6/site-packages/yaml/__init__.py", line 114, in load
    return loader.get_single_data()
  File "/home/robin/Projects/deployment-configs/env3/lib/python3.6/site-packages/yaml/constructor.py", line 41, in get_single_data
    node = self.get_single_node()
  File "/home/robin/Projects/deployment-configs/env3/lib/python3.6/site-packages/yaml/composer.py", line 36, in get_single_node
    document = self.compose_document()
  File "/home/robin/Projects/deployment-configs/env3/lib/python3.6/site-packages/yaml/composer.py", line 58, in compose_document
    self.get_event()
  File "/home/robin/Projects/deployment-configs/env3/lib/python3.6/site-packages/yaml/parser.py", line 118, in get_event
    self.current_event = self.state()
  File "/home/robin/Projects/deployment-configs/env3/lib/python3.6/site-packages/yaml/parser.py", line 193, in parse_document_end
    token = self.peek_token()
  File "/home/robin/Projects/deployment-configs/env3/lib/python3.6/site-packages/yaml/scanner.py", line 129, in peek_token
    self.fetch_more_tokens()
  File "/home/robin/Projects/deployment-configs/env3/lib/python3.6/site-packages/yaml/scanner.py", line 227, in fetch_more_tokens
    return self.fetch_alias()
  File "/home/robin/Projects/deployment-configs/env3/lib/python3.6/site-packages/yaml/scanner.py", line 610, in fetch_alias
    self.tokens.append(self.scan_anchor(AliasToken))
  File "/home/robin/Projects/deployment-configs/env3/lib/python3.6/site-packages/yaml/scanner.py", line 924, in scan_anchor
    % ch, self.get_mark())
yaml.scanner.ScannerError: while scanning an alias
  in "sites/README.md", line 6, column 1
expected alphabetic or numeric character, but found '*'
  in "sites/README.md", line 6, column 2
releases.demo.haus.yaml
sdrsatcom.snapcraft.io.yaml
snapcraft.io.yaml
staging-api.snapcraft.io.yaml
Traceback (most recent call last):
  File "./konf.py", line 157, in <module>
    print(projectConfig.render())
  File "./konf.py", line 121, in render
    deployment_env=self.deployment_env,
  File "/home/robin/Projects/deployment-configs/env3/lib/python3.6/site-packages/jinja2/asyncsupport.py", line 76, in render
    return original_render(self, *args, **kwargs)
  File "/home/robin/Projects/deployment-configs/env3/lib/python3.6/site-packages/jinja2/environment.py", line 1008, in render
    return self.environment.handle_exception(exc_info, True)
  File "/home/robin/Projects/deployment-configs/env3/lib/python3.6/site-packages/jinja2/environment.py", line 780, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "/home/robin/Projects/deployment-configs/env3/lib/python3.6/site-packages/jinja2/_compat.py", line 37, in reraise
    raise value.with_traceback(tb)
  File "./templates/site.yaml", line 15, in top-level template code
    {% include 'deployment.yaml' %}
  File "./templates/deployment.yaml", line 21, in top-level template code
    containerPort: {{ data.containerPort | default(80) }}
jinja2.exceptions.UndefinedError: 'dict object' has no attribute 'replicas'
tutorials.ubuntu.com.yaml
ubuntu.com.yaml
Traceback (most recent call last):
  File "./konf.py", line 157, in <module>
    print(projectConfig.render())
  File "./konf.py", line 121, in render
    deployment_env=self.deployment_env,
  File "/home/robin/Projects/deployment-configs/env3/lib/python3.6/site-packages/jinja2/asyncsupport.py", line 76, in render
    return original_render(self, *args, **kwargs)
  File "/home/robin/Projects/deployment-configs/env3/lib/python3.6/site-packages/jinja2/environment.py", line 1008, in render
    return self.environment.handle_exception(exc_info, True)
  File "/home/robin/Projects/deployment-configs/env3/lib/python3.6/site-packages/jinja2/environment.py", line 780, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "/home/robin/Projects/deployment-configs/env3/lib/python3.6/site-packages/jinja2/_compat.py", line 37, in reraise
    raise value.with_traceback(tb)
  File "./templates/site.yaml", line 24, in top-level template code
    {%- include 'deployment.yaml' %}
  File "./templates/deployment.yaml", line 21, in top-level template code
    containerPort: {{ data.containerPort | default(80) }}
jinja2.exceptions.UndefinedError: 'dict object' has no attribute 'replicas'
usn.ubuntu.com.yaml
vanillaframework.io.yaml
webteam-ircbot.canonical.com.yaml
Traceback (most recent call last):
  File "./konf.py", line 157, in <module>
    print(projectConfig.render())
  File "./konf.py", line 121, in render
    deployment_env=self.deployment_env,
  File "/home/robin/Projects/deployment-configs/env3/lib/python3.6/site-packages/jinja2/asyncsupport.py", line 76, in render
    return original_render(self, *args, **kwargs)
  File "/home/robin/Projects/deployment-configs/env3/lib/python3.6/site-packages/jinja2/environment.py", line 1008, in render
    return self.environment.handle_exception(exc_info, True)
  File "/home/robin/Projects/deployment-configs/env3/lib/python3.6/site-packages/jinja2/environment.py", line 780, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "/home/robin/Projects/deployment-configs/env3/lib/python3.6/site-packages/jinja2/_compat.py", line 37, in reraise
    raise value.with_traceback(tb)
  File "./templates/site.yaml", line 29, in top-level template code
    {% include 'ingress.yaml' %}
  File "./templates/ingress.yaml", line 10, in top-level template code
    {%- if data.nginxConfigurationSnippet %}
jinja2.exceptions.UndefinedError: 'dict object' has no attribute 'nginxConfigurationSnippet'
```

File: sites/staging-api.snapcraft.io.yaml (hunk 3 above)
Comment (nottrobin) [reply sites/staging-api.snapcraft.io.yaml]: :+1:

**Comment by jkfran** [comment]
@nottrobin I updated the branch. Now only should fail "staging-api.snapcraft.io" becuase we don't have production for it.

**Pull request merged**